src/services/chain_spec_service.ts

```diff
@@ -0,0 +1,111 @@
+  bootNodes: string[];
+  name: string;
+  id: string;
+} & Record<string, unknown>;
+
+export type ChainSpecsFetchingState =
+  | { tag: "success"; specs: ChainSpec[] }
+  | { tag: "loading" }
+  | { tag: "error"; error: string };
+
+export class ChainSpecService {
+  chainSpecs: Accessor<ChainSpecsFetchingState>;
+  #setChainSpecs: Setter<ChainSpecsFetchingState>;
+
+  static #instance: ChainSpecService;
+  static get instance(): ChainSpecService {
+    if (ChainSpecService.#instance === undefined) {
+      ChainSpecService.#instance = new ChainSpecService();
+    }
+    return ChainSpecService.#instance;
+  }
+
+  constructor() {
+    const [chainSpecs, setChainSpecs] = createSignal<ChainSpecsFetchingState>({
+      tag: "loading",
+    });
+    this.chainSpecs = chainSpecs;
+    this.#setChainSpecs = setChainSpecs;
+  }
+
+  async fetchAndCacheChainSpecs(): Promise<void> {
+    console.log("ChainSpecService: start fetching chain specs...");
+    this.#setChainSpecs({ tag: "loading" });
+    try {
+      const res = await fetch(SUBSTRATE_CONNECT_CHAIN_SPECS_GITHUB_URL);
+      const json: any = await res.json();
+
+      const specs: ChainSpec[] = [];
+      for (const fileEntry of json) {
+        if (!fileEntry.name.endsWith(".json")) {
+          continue;
+        }
+        const res = await fetch(fileEntry.download_url);
+        const spec: unknown = await res.json();
+        specs.push(spec as ChainSpec);
+      }
+
+      console.log("ChainSpecService: fetching successful!", specs);
+      this.#handleOnLoadCallbacks(specs);
+      this.#setChainSpecs({ tag: "success", specs });
+    } catch (ex: any) {
+      console.error("ChainSpecService: fetching error!", ex);
+      this.#setChainSpecs({
+        tag: "error",
+        error: ex.toString(),
+      });
+    }
+  }
+
+  #on_load_call_backs: ChainSpecResolver[] = [];
+  #handleOnLoadCallbacks(specs: ChainSpec[]) {
+    for (const [chain_name, resolve, reject] of this.#on_load_call_backs) {
+      const spec = specs.find((spec) => spec.name === chain_name);
+      if (spec) {
+        resolve(spec);
+      } else {
+        reject(
+          new Error(
+            `Chain spec with name ${chain_name} not found in loaded presets`
+          )
+        );
+      }
+    }
+    this.#on_load_call_backs = [];
+  }
+
+  loadChainSpecForChainName(chain_name: string): Promise<ChainSpec> {
+    const chainSpecs = this.chainSpecs();
+    switch (chainSpecs.tag) {
+      case "success": {
+        const spec = chainSpecs.specs.find((spec) => spec.name === chain_name);
+        if (spec) {
+          return Promise.resolve(spec);
+        } else {
+          return Promise.reject(
+            new Error(
+              `Chain spec with name ${chain_name} not found in loaded presets`
+            )
+          );
+        }
+      }
+      case "loading": {
+        return new Promise((resolve, reject) => {
+          this.#on_load_call_backs.push([chain_name, resolve, reject]);
```

Comment:
Ah, I see now; you basically have to call `fetchAsnCacheChainSpecs` at least once, and then if this is called during "loading", it's fine because it'll push to a list that hasn't finished iterating yet, and otherwise it'll success or error anyway.

It's the sort of thing that could be simplified by having eg

```
function loadChainSpecs(): Promise<ChainSpecs> { // load chain specs }

let chainSpecs = await loadChainSpecs();

let spec = chainSpecs.get("polkadot") // this is now synchronous and will fail or succeed immediately
```

Reply:
Hmmm I don't know if this approach would work, since we do not want to block until `await loadChainSpecs()` finishes. E.g. I would not want to halt rendering until this is done. And I want to guarantee that the loading stuff only happens exactly once, so `loadChainSpecs` should only be called in one place. I could also kick it off when the `ChainSpecService` singleton is first accessed. I needed some mechanism to have `loadChainSpecForChainName()` be callable from anywhere inside the app any number of times and returning something as soon as the single `loadChainSpecs()` call finishes executing. Not sure if this is the best solution though...

Reply:
I think for now it's all ok; my guess at what I'd do would be to call `loadChainSpecs` or the equivalent when the user first clicks on the "Light Client" option (and show a loading spinner or whatever) and then render the list of available chains we see as a result of that as the radio buttons (possibly filtered to only be ones we are looking for).

The advantage of having the `awaits` all being explicit (aside from simpler code) is that it's up to you whether to wait while it loads or whatever because it's all quite explicit (rather than the loading being hidden behind the scenes). The UI _could_ choose to start loading the specs behind the scenes anyway still.

And if I were to do it I would still probably cache the `loadChainSpecs` result, which you could do quite easily with a little helper eg:

```ts
// Take a function that returns a promise and return a function that
// will only call that once and then cache the result. If the promise
// fails, the cache is cleared.
function cachedPromiseFn<T>(fn: () => Promise<T>): () => Promise<T> {
  let loading: Promise<T> | undefined = undefined;

  return () => {
    if (!loading) {
      loading = fn().catch((e) => {
        loading = undefined;
        throw e
      })
    }
    return loading
  }
}

let loadChainSpecs = cachedPromiseFn(() => {
    // async logic to load the specs
})
```

But yeah, it's all good :)

Reply:
Ahh, that makes so much sense! Storing the fetching Promise itself in a variable like this did not come to mind to me.